Hazelnut/src/EditorLayer.cpp

```diff
@@ -150,14 +159,10 @@ namespace Hazel {
 
 		ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, ImVec2{ 0, 0 });
 		ImGui::Begin("Viewport");
+
 		ImVec2 viewportPanelSize = ImGui::GetContentRegionAvail();
-		if (m_ViewportSize != *((glm::vec2*)&viewportPanelSize))
-		{
-			m_Framebuffer->Resize((uint32_t)viewportPanelSize.x, (uint32_t)viewportPanelSize.y);
-			m_ViewportSize = { viewportPanelSize.x, viewportPanelSize.y };
+		m_ViewportSize = { viewportPanelSize.x, viewportPanelSize.y };
 
-			m_CameraController.OnResize(viewportPanelSize.x, viewportPanelSize.y);
-		}
 		uint32_t textureID = m_Framebuffer->GetColorAttachmentRendererID();
 		ImGui::Image((void*)textureID, ImVec2{ m_ViewportSize.x, m_ViewportSize.y }, ImVec2{ 0, 1 }, ImVec2{ 1, 0 });
 		ImGui::End();
```

Comment:
> The PR as is was to maintain original resize 'flow'.
> 
> Before:
> 
> ```
> OnUpdate
> OnImGuiRender <- Resize Happened in here
> OnUpdate
> ```
> 
> After:
> 
> ```
> OnUpdate
> OnImGuiRender <- Resize change 'request' here
> OnUpdate <- Change viewport and camera before actual render
> ```

I see that you want to maintain the current flow, which is fine by me. After having another look, what you describe is more or less correct, but I think the current solution is not solving the flickering? Let my try and explain..

So first of all you got ImGui that is registering the resizing when you resize the window. This will propagate the new size inside ImGui, so to `ImGui::GetContentRegionAvail();`, which we can then use.

The issue that we're having is that the created framebuffer is not the same dimensions as the one that ImGui is rendering, which is why it's turning black in the first place? The solution for that is to render the 'old' sized framebuffer, like when we hardcoded the checkerboard size before:
- When you make it smaller than the checkerboard size, it will render the excess checkerboard outside of the window, so it won't be visible.
- When you make it bigger, the (black) ImGui background will be visible on the side/bottom.

However I just noticed that you did it in the wrong order:
https://github.com/TheCherno/Hazel/blob/24990184428c4a0e96f9a752df49f00459f79766/Hazelnut/src/EditorLayer.cpp#L160-L169

1. Line 163 you get the new content  size (of the empty window)
2. Line 164 you update the viewport size, which act as the 'request' to update the viewport
3. Line 167 you render the old size viewport, but you render it with the new viewport size.

Where I think you should:

1. Line 163, extract the new content size (of the empty window)
2. Line 167 render the old size viewport with the old viewport size
3. Line 164 update the viewport size, which act as the 'request' to update the viewport.

```suggestion
		ImGui::Begin("Viewport");
		ImVec2 viewportPanelSize = ImGui::GetContentRegionAvail();

		uint32_t textureID = m_Framebuffer->GetColorAttachmentRendererID();
		ImGui::Image((void*)textureID, ImVec2{ m_ViewportSize.x, m_ViewportSize.y }, ImVec2{ 0, 1 }, ImVec2{ 1, 0 });
		
		m_ViewportSize = { viewportPanelSize.x, viewportPanelSize.y };
		ImGui::End();
```

Reply:
Size is unrelated to it turning black. The reason it is flickers black on resize is that on every size change it was creating a new framebuffer texture and directly rendering that in ImGui before actually rendering anything to the framebuffer, aka rendering empty new texture.

Using a different size than the texture in `ImGui::Image` 'only' causes a slight stretch, no flicker really, which I didn't really think about when I created the PR, so I guess something along the lines of your suggestion does make sense.
Though I think instead of using `m_ViewportSize` for the Image render it might be better to use the actual viewport size to avoid first frame incorrect size (0 in this case).

Reply:
> Size is unrelated to it turning black. The reason it is flickers black on resize is that on every size change it was creating a new framebuffer texture and directly rendering that in ImGui before actually rendering anything to the framebuffer, aka rendering empty new texture.
> 
> Using a different size than the texture in `ImGui::Image` 'only' causes a slight stretch, no flicker really, which I didn't really think about when I created the PR, so I guess something along the lines of your suggestion does make sense.
> Though I think instead of using `m_ViewportSize` for the Image render it might be better to use the actual viewport size to avoid first frame incorrect size (0 in this case).

Hmm, I was just playing with this.
Using the old size results in small-ish black bars around the Viewport during resize as well as ImGui flickering the vertical scrollbar on shrinking.
The current PR just stretches the image to always be the same size as the Viewport window.

Personally I am not sure which one I find better.

More correct is the black bars, the flickering scrollbar really makes this not nice and I don't think there is a way in ImGui to disable that at the moment.

Reply:
After some searching: According to ocornut/imgui#1089, there should be a windowflag `ImGuiWindowFlags_NoScrollbar` to disable the scrollbar? Maybe that might help?

Not sure what behaviour we want. My initial thought process was, assume that we stretch the right side, while dragging the screen would jitter a bit. The scroll bar seems a big nono for sure. 

If that windowflag works, we still have to decide which to choose, otherwise the choice is easy to just stretch it a bit... My personal preference would go to the small black boarder when making it larger, if it works, as that will result in 'perfect' behaviour.

Reply:
Ah yes, must have that flag. I does work as advertised, no scrollbars anymore.
So now just need a decision on what we want.

Reply:
Assuming 99.95% of the resizing will happen with the mouse, the keyboard inputs do not affect what's visual on the screen as feedback to application behavior. So we can assume that there are no mouse inputs at the time like clicking swiping, ...

Hence why I would go for the stretching, without setting the WindowFlags, and as a result not have the black borders during that frame, and have a small distortion in the rendering as it stretches/squashes the frame a bit.

Reply:
After testing it on my 4K monitor, I can almost not see the streching/squashing, however the black boarder is alot more obvious, so your approach seems to give the best compromise/result